tests/src/lib.rs

```diff
@@ -54,10 +54,26 @@ pub fn fail<T>(
 }
 
 /// Run some code in an error scope and assert that validation succeeds.
+#[track_caller]
 pub fn valid<T>(device: &wgpu::Device, callback: impl FnOnce() -> T) -> T {
     device.push_error_scope(wgpu::ErrorFilter::Validation);
     let result = callback();
-    assert!(pollster::block_on(device.pop_error_scope()).is_none());
+    if let Some(error) = pollster::block_on(device.pop_error_scope()) {
+        use std::fmt::Write;
+        let mut message = String::new();
+        writeln!(&mut message,
+                 "`valid` block at {} encountered wgpu error",
+                 std::panic::Location::caller()).unwrap();
+        let mut error: &dyn std::error::Error = &error;
+        loop {
+            writeln!(&mut message, "{error}").unwrap();
+            let Some(source) = error.source() else {
+                break;
+            };
+            error = source;
+        }
```

Comment:
To be very concrete with @teoxoy's point, the OP's sample output text illustrates this redundancy:

```
Caused by:
  In Device::create_bind_group_layout, label = 'buffer_storage'
    Too many bindings of type StorageBuffers in Stage ShaderStages(COMPUTE), limit is 0, count was 2. Check the limit `max_storage_buffers_per_shader_stage` passed to `Adapter::request_device`

In Device::create_bind_group_layout, label = 'buffer_storage'
Too many bindings of type StorageBuffers in Stage ShaderStages(COMPUTE), limit is 0, count was 2. Check the limit `max_storage_buffers_per_shader_stage` passed to `Adapter::request_device`
```

…where the top paragraph is the `Display` implementation, and the bottom is the output of this source chain printing loop.

Reply:
Removing the printing loop with 4b1730ffa9139740d096da303db2fc3ba470050b.

Reply:
This is, IMO, backwards — `Display` implementations _shouldn't_ be printing the `source()` chain, precisely so that generic error printers can produce non-redundant text.

Reply:
The reason we do it this way is because we need to support a tree of errors not just a chain. That's what we use the `MultiError` type for.